src/wallet/test/fuzz/wallet_bdb_parser.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+#include <test/util/setup_common.h>
+#include <util/fs.h>
+#include <util/time.h>
+#include <util/translation.h>
+#include <wallet/db.h>
+#include <wallet/dump.h>
+#include <wallet/migrate.h>
+
+#include <iostream>
+
+using wallet::DatabaseOptions;
+using wallet::DatabaseStatus;
+
+namespace {
+TestingSetup* g_setup;
+} // namespace
+
+void initialize_wallet_bdb_parser()
+{
+    static auto testing_setup = MakeNoLogFileContext<TestingSetup>();
+    g_setup = testing_setup.get();
+}
+
+FUZZ_TARGET(wallet_bdb_parser, .init = initialize_wallet_bdb_parser)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+
+    const auto wallet_path = g_setup->m_args.GetDataDirNet() / "fuzzed_wallet.dat";
+
+    {
+        AutoFile outfile{fsbridge::fopen(wallet_path, "wb")};
+        const auto file_data{ConsumeRandomLengthByteVector(fuzzed_data_provider)};
+        outfile << Span{file_data};
+    }
+
+    const DatabaseOptions options{};
+    DatabaseStatus status;
+    bilingual_str error;
+
+    fs::path dumpfile{g_setup->m_args.GetDataDirNet() / "fuzzed_dumpfile.dat"};
+    if (fs::exists(dumpfile)) { // Writing into an existing dump file will throw an exception
+        remove(dumpfile);
+    }
+    g_setup->m_args.ForceSetArg("-dumpfile", fs::PathToString(dumpfile));
+
+    try {
+        auto db{MakeBerkeleyRODatabase(wallet_path, options, status, error)};
+        assert(DumpWallet(g_setup->m_args, *db, error));
```

Comment:
It would be useful here - if BDB is compiled - to also open it with BDB, dump and then compare both dumps.

I'm not sure what to do for cases where BDB and RO have slightly different opinions on what they're willing to open at all. It's probably fine to skip comparing files where _either_ refuses to open.

Reply:
Added a comparison

Reply:
Nice! I'll run the fuzzer against that...